### ISSUE TITLE
If the model is nested, keras.backend.gradients returns None.

### ISSUE THREAD

**Issue by keisen** [keras-team/keras#10850]
- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

---

I'm creating grad-cam and smooth-grad libraries with keras.
keras.backend.gradients returns None, if the model is nested.
I want to avoid this issue.

This issue is occurred when root-model has pre-layer or post-layer of sub-model.
The implementation to reproduce the problem is as follows.

```
import keras.backend as K
from keras.models import Sequential
from keras.layers import Conv2D, Dense, Input, Flatten, Lambda

# Actually this is keras.applications.XXX
sub_cnn_model = Sequential([
    Conv2D(3, 3, activation='relu', input_shape=(28, 28, 1), name='subconv1'),
    Flatten(),
    Dense(1)
])
sub_cnn_model.summary()

# Any model include keras.applications.XXX
# Actually, `pre` and `post` layer are Conv, Dense or Merge instead of Lambda layer.
model = Sequential([
    Lambda(function=lambda x: x, input_shape=(28,28,1), name='any-pre-processing'),
    sub_cnn_model,
    Lambda(function=lambda x: x, name='any-post-processing'),
])
model.summary()

print('grads = ',
      K.gradients(K.sum(model.output), [sub_cnn_model.get_layer('subconv1').output]))
```

The execution result is as follows.

```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
subconv1 (Conv2D)            (None, 26, 26, 3)         30        
_________________________________________________________________
flatten_6 (Flatten)          (None, 2028)              0         
_________________________________________________________________
dense_8 (Dense)              (None, 1)                 2029      
=================================================================
Total params: 2,059
Trainable params: 2,059
Non-trainable params: 0
_________________________________________________________________
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
any-pre-processing (Lambda)  (None, 28, 28, 1)         0         
_________________________________________________________________
sequential_12 (Sequential)   (None, 1)                 2059      
_________________________________________________________________
any-post-processing (Lambda) (None, 1)                 0         
=================================================================
Total params: 2,059
Trainable params: 2,059
Non-trainable params: 0
_________________________________________________________________
grads =  [None]
```

It is a model without problems in learning process, so I think it should be able to get gradients.

Thanks.

**Comment by haimat** [comment]
I ran into the same problem.  
Any news on this issue?

**Comment by quizzy5889** [comment]
Same issue. Is there any update?

**Comment by Ely-S** [comment]
Experiencing this with TF 1.15.0 on TPU v3. Are there any updates?

**Comment by FlavrFlave** [comment]
Hi, I experience the same using 2.3.0-tf :( can anyone provide an update or workaround? Much appreciated, thanks!

**Comment by wingman-jr-addon** [comment]
Well, I did a hacky workaround that worked for me but I do not think it will work for a lot of others. Basically I had a submodel that had some extra layers after it, and I wanted to do GradCAM-type stuff after I had loaded the saved .h5 model. So what I did was find the submodel (which was first) and surgically appended the later layers to the output of the submodel directly, essentially turning the submodel into the main model. Pretty bad but allowed me to do CAM. Note this would have to change if your submodel came in the middle or you had divergent paths, etc. etc.

My original full model looked like this:
```
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to
==================================================================================================
input (InputLayer)              [(None, 224, 224, 3) 0
__________________________________________________________________________________________________
submodel (Model (None, 28, 28, 1280) 2257984     input[0][0]
__________________________________________________________________________________________________
spatial_dropout2d (SpatialDropo (None, 28, 28, 1280) 0           submodel[1][0]
__________________________________________________________________________________________________
final_avg_pool (GlobalAveragePo (None, 1280)         0           spatial_dropout2d[0][0]
__________________________________________________________________________________________________
fc_final (Dense)                (None, 128)          163968      final_avg_pool[0][0]
__________________________________________________________________________________________________
dropout (Dropout)               (None, 128)          0           fc_final[0][0]
__________________________________________________________________________________________________
classifier (Dense)              (None, 4)            516         dropout[0][0]
__________________________________________________________________________________________________
flatten (Flatten)               (None, 1280)         0           final_avg_pool[0][0]
==================================================================================================
```
And I did the following surgery:
```
    submodel_index, submodel = find_layer(full_model, 'submodel_name')
    x = submodel.outputs[0]
    for layer_index in range(submodel_index+1, len(full_model.layers)):
        extracted_layer = full_model.layers[layer_index]
        print('Surgically appending layer : '+str(extracted_layer))
        x = extracted_layer(x)
    new_model = Model(inputs=submodel.inputs, outputs=[x])
```
with helper
```
def find_layer(model, layer_name):
    for (i, layer) in enumerate(model.layers):
        if layer.name == layer_name:
            return (i, layer)
```
I have a hunch it just doesn't quite "see" the submodel's output as connecting to the rest of the graph in quite the same way as normal, hence giving `None` as if it were disconnected.
